[WIP] Add Comment to an Issue

AustinGreen opened the pull request:
This PR is nearly ready. It's a bit noisy due to conflicting prettier formatting. I'll add items to the check list as I think of them. Please offer feedback if I'm missing anything.

Closes #2

rwieruch commented:
This looks great so far! 👍  You could implement the optimistic response in another branch to keep this PR small. But it's up to you :)

AustinGreen commented:
Sounds good. I'll just add a success message instead and refactor a couple small things, and this should be ready

AustinGreen commented:
I added a success and error message when the user leaves a comment. I also refactored the `FetchMore` component, so the text just won't display instead of the "no more" message. This was my first time using Apollo, so I'd love to hear feedback if something does not look right.

I can improve on this by adding the optimistic response in another PR.

rwieruch commented:
Pretty great PR @AustinGreen I will merge it :) I will leave a couple of of comments in the PR for you though.

The refactoring for the FetchMore component is reasonable and all the Apollo implementations look good to me 💯 

- [ ] Maybe we should consider to add a [Prettier configuration file](https://prettier.io/docs/en/configuration.html) to the Project in another PR.

Thanks for this PR @AustinGreen 🎉

rwieruch commented on src/AddComment/index.js:
```diff
@@ -0,0 +1,64 @@
+    const { value } = this.state;
+    return (
+      <div>
+        {this.renderMessage()}
```
Usually I extract these things as own components like <RequestMessage /> but keep them still in the same file. However, good to see different approaches! Your way you have direct access to the state object! 👍

AustinGreen replied:
I used a method instead of a separate component because it was not needed for reusability, but I'm sure this a premature optimization. I've also seen static methods used, so it would look like `
```
static renderMessage = ({ displaySuccess, displayError, errorMessage }) => {
  ...
}

render() {
  const { value, displaySuccess, displayError, errorMessage } = this.state;
  return (
      <div>
        <AddComment.renderMessage displaySuccess={displaySuccess} displayError={displayError} errorMessage={errorMessage} /> 
...
```

Curious to hear your thoughts on this.

rwieruch replied:
You are totally right! From a technical point of view it makes no sense to extract it unless you would want to reuse it or test it as standalone component. I guess for me it is more of a stylistic preference to extract atomic problem solvers as their own functions :)

AustinGreen replied:
I first saw it used by @kentcdodds. It gives you the stylistic benefit of code separation without creating another component.

kentcdodds replied:
Woohoo!

kentcdodds replied:
Though truth be told I'd actually prefer to extract that method to be a pure function which accepts what it needs from props/state as an object argument. That way it's really clear what that part of the JSX needs, and it would also be pretty trivial to convert it to a full blown render prop later if desired.

AustinGreen replied:
@kentcdodds thanks for the input! Not that it makes a difference, but do you mean option #1 or #2 below?

```
function MessageBox({ messageState, errorMessage }) {
  return {
    success: <div className="AddComment-message AddComment-message--success">Your comment has been posted</div>,
    error: <div className="AddComment-message AddComment-message--error">{errorMessage}</div>,
    none: null
  }[messageState];
}
...

class AddComment extends Component {

...

const messageState = isSuccess ? 'success' : isError ? 'error' : 'none';
  return (
    <div>
      <MessageBox messageState={messageState} errorMessage={errorMessage} /> <-- Option #1
      {MessageBox({ messageState, errorMessage })} <-- Option #2
```

kentcdodds replied:
Option 2, but I'd lower-case it like a normal function. No need to make it a component necessarily. But really it doesn't matter all that much to me. Pretty loosy-goosy here.

```javascript
function renderMessage({ displaySuccess, displayError, errorMessage }) {
    if (displaySuccess) {
      return <div className="AddComment-message AddComment-message--success">Your comment has been posted</div>;
    } else if (displayError) {
      return <div className="AddComment-message AddComment-message--error">{errorMessage}</div>;
    } else {
      return null;
    }
  }
```

rwieruch commented on src/AddComment/index.js:
```diff
@@ -0,0 +1,64 @@
+
+  renderMessage = () => {
+    const { displaySuccess, displayError, errorMessage } = this.state;
+    if (displaySuccess) {
```
Just in case you didn't know about it: Maybe the [enum conditional rendering](https://www.robinwieruch.de/conditional-rendering-react/#enums) could be used instead of all these if else statements. I guess in this case it would mean that the local state of the component needs to change slightly to be mapped to SUCCESS, ERROR, NONE.
